File: backend/src/main/java/com/finance/dto/RecurringDtos.java
package com.finance.dto;

import com.finance.entities.RecurringFrequency;
import com.finance.entities.TransactionType;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.UUID;

public class RecurringDtos {

    public record RecurringTransactionRequest(
            @NotBlank String title,
            @NotNull TransactionType type,
            @NotNull @DecimalMin(value = "0.01") BigDecimal amount,
            UUID categoryId,
            @NotNull UUID accountId,
            @NotNull RecurringFrequency frequency,
            @NotNull LocalDate startDate,
            LocalDate endDate,
            @NotNull LocalDate nextRunDate,
            boolean autoCreateTransaction
    ) {
    }

    public record RecurringTransactionResponse(
            UUID id,
            String title,
            TransactionType type,
            BigDecimal amount,
            UUID categoryId,
            String categoryName,
            UUID accountId,
            String accountName,
            RecurringFrequency frequency,
            LocalDate startDate,
            LocalDate endDate,
            LocalDate nextRunDate,
            boolean autoCreateTransaction
    ) {
    }
}

File: backend-dotnet/FinanceTracker.Api/Contracts/RecurringDtos.cs
using System.ComponentModel.DataAnnotations;
using FinanceTracker.Api.Entities;
using FinanceTracker.Api.Infrastructure;

namespace FinanceTracker.Api.Contracts;

public sealed record RecurringTransactionRequest(
    [param: NotBlank(ErrorMessage = "must not be blank")]
    string? title,
    [param: Required(ErrorMessage = "must not be null")]
    TransactionType? type,
    [param: Required(ErrorMessage = "must not be null")]
    [param: MinValue("0.01", ErrorMessage = "must be greater than or equal to 0.01")]
    decimal? amount,
    Guid? categoryId,
    [param: Required(ErrorMessage = "must not be null")]
    Guid? accountId,
    [param: Required(ErrorMessage = "must not be null")]
    RecurringFrequency? frequency,
    [param: Required(ErrorMessage = "must not be null")]
    DateOnly? startDate,
    DateOnly? endDate,
    [param: Required(ErrorMessage = "must not be null")]
    DateOnly? nextRunDate,
    bool autoCreateTransaction
);

public sealed record RecurringTransactionResponse(
    Guid id,
    string title,
    TransactionType type,
    decimal amount,
    Guid? categoryId,
    string? categoryName,
    Guid accountId,
    string accountName,
    RecurringFrequency frequency,
    DateOnly startDate,
    DateOnly? endDate,
    DateOnly nextRunDate,
    bool autoCreateTransaction
);
